### ISSUE TITLE
Returning torch object from multiprocessing pipe breaks random generator state

### ISSUE THREAD

**Issue by timmeinhardt** [pytorch/pytorch#8651]
While working with multiprocessing and PyTorch 0.4.0 I encountered a (for me) unintuitive behaviour of the random generator state. Sorry for the lengthy code example but I hope it helps to clarify my problem. Usually the random state of a subprocess is independent of the parent process but my example shows that if a PyTorch object is returned from a subprocess the random state of the same is altered. From my understanding only returning a PyTorch object should not effect the random state in such a way: The randomly generated sequences in `RETURN TORCH OBJECT` and `RETURN NONE` should be identical. Can someone maybe clarify my confusion?

```
import torch
from torch.multiprocessing import Pipe, Process


def worker(remote, parent_remote):
    parent_remote.close()
    while True:
        cmd, return_torch = remote.recv()

        ret = None
        if return_torch:
            ret = torch.Tensor([1, 2, 3])

        if cmd == 'seed':
            torch.random.manual_seed(100)
            remote.send(ret)
        elif cmd == 'print':
            print(torch.randperm(5))
            remote.send(ret)
        elif cmd == 'all':
            torch.random.manual_seed(100)
            print(torch.randperm(5))
            print(torch.randperm(5))
            remote.send(ret)


remote, work_remote = Pipe()
ps = Process(target=worker, args=(work_remote, remote))
ps.daemon = True
ps.start()
work_remote.close()

print("RETURN TORCH OBJECT")

remote.send(['all', True])
remote.recv()

remote.send(['seed', True])
remote.recv()
remote.send(['print', True])
remote.recv()
remote.send(['print', True])
remote.recv()

print("RETURN NONE")

remote.send(['all', False])
remote.recv()

remote.send(['seed', False])
remote.recv()
remote.send(['print', False])
remote.recv()
remote.send(['print', False])
remote.recv()
```

Console output:
```
RETURN TORCH OBJECT
tensor([ 0,  1,  4,  2,  3])
tensor([ 0,  4,  2,  3,  1])
tensor([ 2,  4,  1,  0,  3])
tensor([ 2,  3,  0,  1,  4])
RETURN NONE
tensor([ 0,  1,  4,  2,  3])
tensor([ 0,  4,  2,  3,  1])
tensor([ 0,  1,  4,  2,  3])
tensor([ 0,  4,  2,  3,  1])
```

**Comment by SsnL** [comment]
Hi, could you elaborate on what you expect to see here? From my understanding, since you are not seeding in `'print'` case, it inherits parent rng state, and thus always prints random values. So the output looks expected to me.

**Comment by timmeinhardt** [comment]
I agree that the subprocess inherits the seed but I am resetting the seed of the subprocess with the `seed` or `all` command.
I expect the console output of `RETURN TORCH OBJECT` and `RETURN NONE` to be the same because the only difference is that I return a PyTorch object which should not have an influence on the randomization.

**Comment by SsnL** [comment]
Oh I see. My bad. I somehow thought that you are starting new subprocesses. Yeah this feels unintuitive. Could you try also building the tensor, but just not returning it when `return_torch=False`? It might help locate the issue.

**Comment by timmeinhardt** [comment]
Changing the corresponding lines to
```
ret = None
tensor = torch.Tensor([1, 2, 3])
if return_torch:
    ret = tensor
```
results in the same output as above.

**Comment by yf225** [comment]
@timmeinhardt It seems that the pickling of a `torch.Tensor` by `Connection.send()` would somehow alter the RNG state. We are working on better support of Pipe in `torch.multiprocessing` (https://github.com/pytorch/pytorch/issues/8374) which will address this issue. For now, one workaround is to do the following:
```
import io

...
if return_torch:
    buf = io.BytesIO()
    torch.save(tensor, buf)
    buf.seek(0)
    ret = buf
...
```

Then, we can retrieve the tensor by `tensor = torch.load(remote.recv())`.

**Comment by yf225** [comment]
@timmeinhardt The original issue is fixed on current master as well.
